Bricks/sync/owned_borrowed.h

```diff
@@ -0,0 +1,492 @@
+
+  UniqueInstanceContainer() : movable_instance_(std::make_unique<UniqueInstance<T>>()) {}
+
+  template <typename... ARGS>
+  UniqueInstanceContainer(ARGS&&... args)
+      : movable_instance_(std::make_unique<UniqueInstance<T>>(std::forward<ARGS>(args)...)) {}
+
+  UniqueInstanceContainer(UniqueInstanceContainer&&) = default;
+  UniqueInstanceContainer& operator=(UniqueInstanceContainer&&) = default;
+
+  UniqueInstanceContainer(const UniqueInstanceContainer&) = delete;
+  UniqueInstanceContainer& operator=(const UniqueInstanceContainer&) = delete;
+};
+
+}  // namespace current::sync::impl
+
+template <typename T>
+struct ConstructOwned {};
+
+// `Owned<T>` is constructible, movable, and can spin off `BorrowedWithCallback<T>` offspring.
+template <typename T>
+class Owned final : private impl::UniqueInstanceContainer<T>, public WeakBorrowed<T> {
+ private:
+  static_assert(std::is_same<T, decay<T>>::value, "`Owned<>` requires a clean type.");
+
+  using impl_t = impl::UniqueInstanceContainer<T>;
+  using base_t = WeakBorrowed<T>;
+
+ public:
+  Owned() : impl_t(), base_t(impl::ConstructOwned(), *impl_t::movable_instance_) {}
+
+#ifndef CURRENT_OWNED_BORROWED_EXPLICIT_ONLY_CONSTRUCTION
+  // For the end users of Current, we allow "manual" `Owned` construction via `ARGS...`.
+  // (It's really for "backwards compatibility" reasons, we could revisit this. -- D.K.)
+  // Within Current, `MakeOwned<T>(...)` or `Owned<T>(ConstructOwned<T>(), ...)` are used, and thus
+  // the "liberal" constructor is unnecessary.
+  // Original background/rationale: Easier type-level development/debugging of Current itself.
+  template <typename... ARGS>
+  Owned(ARGS&&... args)
+      : impl_t(std::forward<ARGS>(args)...), base_t(impl::ConstructOwned(), *impl_t::movable_instance_) {}
+#endif  // CURRENT_OWNED_BORROWED_EXPLICIT_ONLY_CONSTRUCTION
+
+  template <typename... ARGS>
+  Owned(ConstructOwned<T>, ARGS&&... args)
+      : impl_t(std::forward<ARGS>(args)...), base_t(impl::ConstructOwned(), *impl_t::movable_instance_) {}
+
+  Owned(Owned&& rhs)
+      : impl_t(impl::ConstructUniqueContainerViaMoveConstructor(), std::move(static_cast<impl_t&&>(rhs))),
+        base_t(impl::ConstructOwned(), *impl_t::movable_instance_) {
+    rhs.base_t::p_actual_instance_ = nullptr;
+  }
+
+  Owned& operator=(Owned&& rhs) {
+    impl_t::operator=(std::move(rhs));
+    base_t::InitializedMovedOwned(*impl_t::movable_instance_);
+    rhs.p_actual_instance_ = nullptr;
+    return *this;
+  }
+
+  Owned(const Owned&) = delete;
+  Owned& operator=(const Owned&) = delete;
+};
+
+// The constructor of `BorrowedWithCallback<T>` requires two parameters:
+// 1) `const WeakBorrowed<T>&`, and
+// 2) a required callback, which will be called once the scope of the `Owned<T>` will be terminating.
+// Obviously, an instance of `BorrowedWithCallback` can't be moved natively, as the
+// destruction callback it uses must point to the respective owner object, not to the "old" one.
```

Comment:
Could you elaborate pls? Callback is just a `std::function<void()>` and doesn't point to anything.